File: src/conduit/loader/parquet_loader.py
"""Parquet destination writer."""

from __future__ import annotations

import logging
from pathlib import Path

import pyarrow as pa
import pyarrow.parquet as pq

from conduit.models import DestinationConfig

logger = logging.getLogger(__name__)


def load_parquet(table: pa.Table, dest: DestinationConfig, base_dir: Path) -> None:
    """Write an Arrow table to a Parquet file."""
    file_path = base_dir / dest.config.get("path", f"{dest.name}.parquet")
    file_path.parent.mkdir(parents=True, exist_ok=True)

    compression = dest.config.get("compression", "snappy")
    row_group_size = dest.config.get("row_group_size", dest.batch_size)

    pq.write_table(
        table,
        str(file_path),
        compression=compression,
        row_group_size=row_group_size,
    )
    logger.info(
        "Loaded %d rows to Parquet destination '%s' (%s)",
        len(table), dest.name, file_path,
    )

File: src/ts/loader/parquet_loader.ts
/**
 * Parquet destination writer using DuckDB COPY TO.
 */

import { join, dirname } from "@std/path";
import { DuckDBInstance } from "@duckdb/node-api";
import type { DataTable, DestinationConfig } from "../models.ts";
import { writeCsvSync } from "../util.ts";

export async function loadParquet(
  table: DataTable,
  dest: DestinationConfig,
  baseDir: string,
): Promise<void> {
  const filePath = join(baseDir, (dest.config.path as string) ?? `${dest.name}.parquet`);
  Deno.mkdirSync(dirname(filePath), { recursive: true });

  const compression = (dest.config.compression as string) ?? "snappy";

  const instance = await DuckDBInstance.create();
  const conn = await instance.connect();
  const tmpDir = Deno.makeTempDirSync();

  try {
    // Write data to temp CSV, load into DuckDB, then COPY TO parquet
    const csvPath = `${tmpDir}/_data.csv`;
    writeCsvSync(csvPath, table);

    await conn.run(
      `CREATE TABLE _data AS SELECT * FROM read_csv('${csvPath}', auto_detect=true)`,
    );
    await conn.run(
      `COPY _data TO '${filePath}' (FORMAT PARQUET, COMPRESSION '${compression}')`,
    );

    console.log(
      `Loaded ${table.rows.length} rows to Parquet destination '${dest.name}' (${filePath})`,
    );
  } finally {
    conn.closeSync();
    try {
      Deno.removeSync(tmpDir, { recursive: true });
    } catch {
      // best effort
    }
  }
}
